### PR TITLE
Write generators only once

### LINKED ISSUE #1680: When using a base Conanfile, settings and generators get duplicated
When using a base conanfile, settings and generators values are getting duplicated. I have this base conanfile:

```python
class BaseConan(ConanFile):
    license = "None"
    description = "A base conanfile for common program options"
    settings = "os", "build_type", "arch", "multiarch", "toolchain"
    generators = "cmake",
```

When I go to use it in another conanfile.py, upload it, and run `conan inspect` on it, I get the following partial output
```
generators: ('cmake', 'cmake')
exports: None
exports_sources: None
short_paths: False
apply_env: True
build_policy: None
revision_mode: scm
settings: ('os', 'build_type', 'arch', 'os', 'build_type', 'arch')
```

The conanfile.py that I'm using it in doesn't declare the `generators` or `settings` variables, it just has the following `init` method:
```python
class DerivedConan(ConanFile):
    python_requires = "base/1.0@user/channel"
    python_requires_extend = "base.BaseConan"

    def init(self):
        base = self.python_requires["base"].module.BaseConan
        self.settings = base.settings + self.settings
        self.generators = base.generators + self.generators
        self.options = dict(base.options, **self.options)
        self.license = base.license
```
Based on the [example](https://docs.conan.io/en/latest/reference/conanfile/methods.html#init), I wouldn't expect `self.settings` of `DerivedConan` to be the same as `base.settings`.

This also begs the question as to whether the underlying type of `settings` and `generators` should be a `set`. I got around the issue by just doing:

```python
    def init(self):
        base = self.python_requires["base"].module.BaseConan
        self.settings = set(base.settings + self.settings)
        self.generators = set(base.generators + self.generators)
        self.options = dict(base.options, **self.options)
        self.license = base.license
```

### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 14.04
  * Conan version: 1.24.1
  * Python version: 3.5

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: omit
Docs: omit

This PR ensures that generators are written only once even if they are repeated

closes https://github.com/conan-io/docs/issues/1680

### CODE REVIEW

File: conans/client/generators/__init__.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def __getitem__(self, key):
 def write_generators(conanfile, path, output):
     """ produces auxiliary files, required to build a project or a package.
     """
-    for generator_name in conanfile.generators:
+    for generator_name in set(conanfile.generators):
```
Comment (memsharded) [inline conans/client/generators/__init__.py:91]:
```suggestion
    for generator_name in sorted(set(conanfile.generators)):
```

Comment (jgsogo) [reply conans/client/generators/__init__.py:91]: It shouldn't be needed, is there any reason why you should care about the order of generators (even for testing)?

**Review by czoido — APPROVED** [review]

File: conans/client/generators/__init__.py (hunk 1 above)
Comment (memsharded) [reply conans/client/generators/__init__.py:91]: Maybe a ``attr("slow")`` test checks the output and it comes in a different order? Or some weird interaction between different generators that try to create the same file (``conan_common.cmake``) in the future, and that fail randomly depending on the order. Not a big thing, just a minor detail, 1 out of 10 of this kind of things come later to bite. So can be left as-is.

Comment (jgsogo) [reply conans/client/generators/__init__.py:91]: If it fails once I'll fix that test, noone should expect generators to be in any specific order.

**Pull request merged**